### DISCUSSION TITLE
How to get sponsor info with their corresponding tier info in 1 query

### DISCUSSION THREAD

**Discussion by RaenonX** [community/community#75454] [Apps, API and Webhooks] [answered]
### Select Topic Area

Question

### Body

I don't have preference in between REST and GQL, but I've been trying using GQL because it seems to have more community support.

I am polling sponsors for every 10 mins to make sure my digital perks linked to my sponsorship have the up-to-date sponsorship status. The perks granted are decided based on the subscription tier, so I need something that identifies both user and their sponsoring tier.

So far, I can only use these to get the user info:
```graphql
query {
  user(login:"raenonx") {
    sponsors(first: 10) {
      totalCount
      nodes {
        ... on User { 
          login
          email
        }
      }
    }
  }
}
```

And then use this to get the sponsorship tier of a sponsor:
```graphql
query {
  user(login:"<MySubscriber>") {
    sponsorshipForViewerAsSponsorable {
      id
      tier {
        id
        name
      }
    }
  }
}
```

However, this means 60 sponsors will need 61 API calls, which I'd like to avoid. Is there any query I can use for getting user info and tier info at once?

**Comment by b8zeek** [comment, marked as answer]
Is this what you're looking for buddy? 🤔 

```
query {
  user(login: "raenonx") {
    sponsorshipsAsMaintainer(first: 10) {
      nodes {
        sponsorEntity {
          ... on User {
            login
            email
          }
        }
        tier {
          id
          name
        }
      }
    }
  }
}
```

Kind regards! 🍺 

🍀

**Reply by RaenonX** [reply]
Probably not? I thought of this but I have no idea why it returns nothing (I have 3 sponsors now already)

![image](https://github.com/community/community/assets/18443743/9be4c4bc-d982-490c-b0b6-937a16fdbdb7)

**Reply by RaenonX** [reply]
I forgot to add that they're private sponsors, so I might have to provide PAT to get my private sponsor, if that's supported? If so, how do i do that?

**Reply by b8zeek** [reply]
The query I wrote above should do the trick. Regarding your sponsors, the fact that they're private matters only in case you don't pass your personal access token which I assume is not your case. Try to authenticate your query and you should get the results.

🍀

**Reply by RaenonX** [reply]
Is there any documentation about how to authorize my query? I think I am missing this part but not sure how to auth it.

**Reply by b8zeek** [reply]
Hey, sorry for the late reply, I am traveling, I am at the hackathon... I hope you solved this, in case you didn't here are some things you might check.

Regarding the authentication, it is pretty straight forward. Here's the documentation, it explains creating a token flow, token types, GraphQL authentication...

https://docs.github.com/en/graphql/guides/forming-calls-with-graphql#authenticating-with-a-personal-access-token

I would suggest testing it by querying for the user, you should get your data back.

Good luck. 🍺 

🍀

**Reply by RaenonX** [reply]
Unfortunately not :( https://github.com/orgs/community/discussions/75454#discussioncomment-7604386

I can't tell what I was missing, hope you had a good travel+hackathon experience

**Reply by b8zeek** [reply]
Hey, the hackathon was amazing, great experience, we came unprepared, still created something amazing and won some great prizes. 🙏

I just tested the query from above and this is the result I got:

```
{
  "data": {
    "user": {
      "sponsorshipsAsMaintainer": {
        "nodes": [
          {
            "sponsorEntity": {
              "login": "DubiousAlien",
              "email": ""
            },
            "tier": null
          }
        ]
      }
    }
  }
}
```

You are still getting the empty list? I would assume you aren't properly authenticated because the query obviously works. Let me know what is happening.

🍀

**Reply by RaenonX** [reply]
Good to hear you won something!

![image](https://github.com/community/community/assets/18443743/38ffae17-8a44-4adb-a10f-a891d95791c1)

So this is what I have, it does have something, but I am not seeing private sponsors. Do you happen to know what token permission will make it work and how should I put it to [GQL explorer](https://docs.github.com/en/graphql/overview/explorer)?

Having this in "Headers" doesn't help:

```json
{
  "Authentication": "Bearer <PAT_WITH_ALL_PERMISSIONS>"
}
```

**Reply by b8zeek** [reply]
When you try authenticating GQL Explorer, I think basic setup involves what you need as well, and that is `Personal user data`. You are offered to grant or request access to the organizations below but you don't need it. I am not sure if you're using the same flow, open GQL Explorer and just go to `Sign In`, follow the steps.

Maybe you don't have any more sponsors... 🤷🏼‍♂️ 

🍀

**Reply by RaenonX** [reply]
Hmm in that case I did sign in.

![image](https://github.com/community/community/assets/18443743/1790a7ea-1094-4bc6-a2e5-4b39ba7dcce4)

![image](https://github.com/community/community/assets/18443743/c0f53642-dbc2-4ba0-b633-5c06a5ff7e6b)

**Reply by RaenonX** [reply]
Turns out this is what it needs to be - `includePrivate: true` has to be there.

```graphql
query {
  user(login: "raenonx") {
    sponsorshipsAsMaintainer(first: 10, includePrivate: true) {
      nodes {
        sponsorEntity {
          ... on User {
            login
            email
          }
        }
        tier {
          id
          name
        }
      }
    }
  }
}
```

**Reply by b8zeek** [reply]
Amazing! I am glad you solved it, great job. 🚀 

P.S. Thanks for the update Mister. 🙏

Best of luck going forward.

🍀

**Comment by RaenonX** [comment]
![image](https://github.com/community/community/assets/18443743/dd5cbf24-296a-4b1e-9aca-ffd39e2eb9e9)

Not sure what went wrong, but this query with the token below still showing nothing - token is classic and has all permissions (I don't want to set expiry on it)

**Comment by RaenonX** [comment]
bump
